compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -486,8 +487,8 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
 
         // FIXME: Add check for trait bound that is already present, particularly `?Sized` so we
         //        don't suggest `T: Sized + ?Sized`.
-        let mut hir_id = body_id;
-        while let Some(node) = self.tcx.hir().find(hir_id) {
+        let mut hir_id = self.tcx.hir().local_def_id_to_hir_id(body_id);
+        while let Some(node) = self.tcx.hir().find_by_def_id(body_id) {
```

Comment:
this will loop forever

Reply:
you need to remove `mut hir_id`, and then at the bottom of the loop reassign `body_id = tcx.local_parent(body_id);` instead so you're actually going up the def id tree.

Reply:
I think you mean the following refactoring?

```rust
while let Some(node) = self.tcx.local_parent(body_id) {
```

Reply:
No, you still need `while let Some(node) = self.tcx.hir().find_by_def_id(body_id) { .. }`.

Line 669 will become:

`body_id = self.tcx.local_parent(body_id);`

And you will delete 490, and make `body_id` on 477 mutable.

Reply:
Ah, ok I think I get it, this while looks stau stuck because the `body_id` never change! thanks to catch it.

However, I did not understand this

>  instead so you're actually going up the def id tree.

Reply:
I think this diff make sense

```diff
diff --git a/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs b/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
index 4c77537a0a5..0ada543b731 100644
--- a/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
@@ -487,8 +487,9 @@ fn suggest_restricting_param_bound(
 
         // FIXME: Add check for trait bound that is already present, particularly `?Sized` so we
         //        don't suggest `T: Sized + ?Sized`.
-        let mut hir_id = self.tcx.hir().local_def_id_to_hir_id(body_id);
+        let mut body_id = body_id;
         while let Some(node) = self.tcx.hir().find_by_def_id(body_id) {
+            let hir_id = self.tcx.hir().local_def_id_to_hir_id(body_id);
             match node {
                 hir::Node::Item(hir::Item {
                     ident,
@@ -665,8 +666,7 @@ fn suggest_restricting_param_bound(
 
                 _ => {}
             }
-
-            hir_id = self.tcx.hir().get_parent_item(hir_id).into();
+           body_id = self.tcx.local_parent(body_id);
         }
     }

```